src/hashset.c

```diff
@@ -0,0 +1,1681 @@
+ * the bucket containing the matching element and points 'pos_in_bucket' to the
+ * index within the bucket. Returns NULL if no matching element was found.
+ *
+ * If 'table_index' is provided, it is set to the index of the table (0 or 1)
+ * the returned bucket belongs to. */
+static bucket *findBucket(hashset *s, uint64_t hash, const void *key, int *pos_in_bucket, int *table_index) {
+    if (hashsetSize(s) == 0) return 0;
+    uint8_t h2 = highBits(hash);
+    int table;
+
+    /* Do some incremental rehashing. */
+    rehashStepOnReadIfNeeded(s);
+
+    /* Check rehashing destination table first, since it is newer and typically
+     * has less 'everfull' flagged buckets. Therefore it needs less probing for
+     * lookup. */
+    for (table = 1; table >= 0; table--) {
+        if (s->used[table] == 0) continue;
+        size_t mask = expToMask(s->bucketExp[table]);
+        size_t bucket_idx = hash & mask;
+        size_t start_bucket_idx = bucket_idx;
+        while (1) {
+            bucket *b = &s->tables[table][bucket_idx];
+            /* Find candidate elements with presence flag set and matching h2 hash. */
+            for (int pos = 0; pos < ELEMENTS_PER_BUCKET; pos++) {
+                if ((b->presence & (1 << pos)) && b->hashes[pos] == h2) {
+                    /* It's a candidate. */
+                    void *element = b->elements[pos];
+                    const void *elem_key = elementGetKey(s, element);
+                    if (compareKeys(s, key, elem_key) == 0) {
+                        /* It's a match. */
+                        if (pos_in_bucket) *pos_in_bucket = pos;
+                        if (table_index) *table_index = table;
+                        return b;
+                    }
+                }
+            }
+
+            /* Probe the next bucket? */
+            if (!b->everfull) break;
+            bucket_idx = nextCursor(bucket_idx, mask);
+            if (bucket_idx == start_bucket_idx) {
+                /* We probed the whole table. This should be extremely rare but
+                 * theoretically it can happen. */
+                break;
+            }
+        }
+    }
+    return NULL;
+}
+
+/* Find an empty position in the table for inserting an element with the given hash. */
+static bucket *findBucketForInsert(hashset *s, uint64_t hash, int *pos_in_bucket, int *table_index) {
+    int table = hashsetIsRehashing(s) ? 1 : 0;
+    assert(s->tables[table]);
+    size_t mask = expToMask(s->bucketExp[table]);
+    size_t bucket_idx = hash & mask;
+    while (1) {
+        bucket *b = &s->tables[table][bucket_idx];
+        for (int pos = 0; pos < ELEMENTS_PER_BUCKET; pos++) {
+            if (b->presence & (1 << pos)) continue; /* busy */
+            if (pos_in_bucket) *pos_in_bucket = pos;
+            if (table_index) *table_index = table;
+            return b;
+        }
+        bucket_idx = nextCursor(bucket_idx, mask);
```

Comment:
Yes! At least there was until I added rehashing when it gets too full. There's a unit test case for that.

Reply:
Well, scan got into an infinite loop, but it's a very similar condition.

Reply:
Yeah, I forgot to delete this comment (like I said I should) but I commented on the insertion path which doesn't seem to have a guard.

Should we have some assertion here or breakout condition here just in case?

Reply:
Do you mean store `start_bucket` before the loop and assert that we never get back to the start bucket?